lib/delta-chain/src/Data/Table.hs

```diff
@@ -0,0 +1,298 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Table
+    ( -- * Synopsis
+
+      -- | 'Table' models a database table.
+      -- It corresponds to a collection of rows.
+      -- Each row has a unique ID, but this is transparent to the API user.
+      --
+      -- 'Pile' models a set of values.
+      -- Unlike 'Set', it is represented as a lightweight list.
+      -- This is used to highlight that the ordering of rows
+      -- in a 'Table' is /not/ deterministic.
+      --
+      -- 'Supply' is a supply of unique IDs.
+
+      -- * Table
+      Table (..)
+    , empty
+    , fromRows
+    , fromList
+    , toPile
+    , toRows
+    , selectWhere
+    , insertMany
+    , deleteWhere
+    , updateWhere
+    , DeltaTable (..)
+    , DeltaDB (..)
+    , tableIntoDatabase
+
+      -- * Pile
+    , Pile (..)
+    , fromSet
+    , deltaListToPile
+    , deltaListFromPile
+    , deltaSetToPile
+    , deltaSetFromPile
+
+      -- * Supply
+    , Supply
+    , abundance
+    , fresh
+    , consume
+    ) where
+
+import Prelude
+
+import Control.Monad
+    ( forM
+    )
+import Control.Monad.Trans.State.Strict
+    ( evalState
+    , state
+    )
+import Data.Delta
+    ( Delta (..)
+    , DeltaList (..)
+    , DeltaSet
+    , DeltaSet1 (..)
+    , Embedding
+    , Embedding' (..)
+    , mkEmbedding
+    )
+import Data.IntMap.Strict
+    ( IntMap
+    )
+import Data.List
+    ( sort
+    , sortOn
+    )
+import Data.Ord
+    ( Down (..)
+    )
+import Data.Set
+    ( Set
+    )
+
+import qualified Data.Delta as Delta
+import qualified Data.IntMap.Strict as Map
+import qualified Data.Set as Set
+
+{-------------------------------------------------------------------------------
+    Table
+-------------------------------------------------------------------------------}
+
+-- | A 'Table' is a collection of rows.
+data Table row = Table
+    { rows :: IntMap row
+    -- ^ Rows indexed by unique ID.
+    , uids :: Supply
+    -- ^ Unique ID supply.
+    -- WARNING: This is an internal part of the structure.
```

Comment:
if so what about not persuading users to construct `Table` like above but add smart constructors/pattern synonyms hiding `uids` ?

Reply:
Yes. At the moment, the module `Data.Table` is preliminary, and will be adjusted later. I think that I had exposed the internal constructor in order to implement reading / writing `Data.Table` from an SQLite database table.